server/sonar-server/src/main/java/org/sonar/server/test/ws/TestsListAction.java

```diff
@@ -0,0 +1,221 @@
+ */
+
+package org.sonar.server.test.ws;
+
+import com.google.common.base.Function;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.io.Resources;
+import org.sonar.api.server.ws.Request;
+import org.sonar.api.server.ws.Response;
+import org.sonar.api.server.ws.WebService;
+import org.sonar.api.utils.text.JsonWriter;
+import org.sonar.api.web.UserRole;
+import org.sonar.core.component.ComponentDto;
+import org.sonar.core.persistence.DbSession;
+import org.sonar.core.persistence.MyBatis;
+import org.sonar.core.util.NonNullInputFunction;
+import org.sonar.server.db.DbClient;
+import org.sonar.server.es.SearchOptions;
+import org.sonar.server.es.SearchResult;
+import org.sonar.server.test.index.CoveredFileDoc;
+import org.sonar.server.test.index.TestDoc;
+import org.sonar.server.test.index.TestIndex;
+import org.sonar.server.user.UserSession;
+
+import javax.annotation.Nullable;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static org.elasticsearch.common.base.MoreObjects.firstNonNull;
+
+public class TestsListAction implements TestAction {
+  public static final String TEST_UUID = "testUuid";
+  public static final String TEST_FILE_UUID = "testFileUuid";
+  public static final String TEST_FILE_KEY = "testFileKey";
+  public static final String SOURCE_FILE_UUID = "sourceFileUuid";
+  public static final String SOURCE_FILE_LINE_NUMBER = "sourceFileLineNumber";
+
+  private final DbClient dbClient;
+  private final TestIndex testIndex;
+
+  public TestsListAction(DbClient dbClient, TestIndex testIndex) {
+    this.dbClient = dbClient;
+    this.testIndex = testIndex;
+  }
+
+  @Override
+  public void define(WebService.NewController controller) {
+    WebService.NewAction action = controller
+      .createAction("list")
+      .setDescription("Get the list of tests.<br /> " +
+        "Require Browse permission on file's project.<br /> " +
+        "One of the following combination of fields must be filled: " +
+        "<ul>" +
+        "<li>Test file UUID</li>" +
+        "<li>Test UUID</li>" +
+        "<li>Source file UUID and Source file line number</li>" +
+        "</ul>")
+      .setSince("5.2")
+      .setResponseExample(Resources.getResource(getClass(), "tests-example-list.json"))
+      .setHandler(this)
+      .addPagingParams(100);
+
+    action
+      .createParam(TEST_FILE_UUID)
+      .setDescription("Test file UUID")
+      .setExampleValue("ce4c03d6-430f-40a9-b777-ad877c00aa4d");
+
+    action
+      .createParam(TEST_FILE_KEY)
+      .setDescription("Test file key")
+      .setExampleValue("org.codehaus.sonar:sonar-server:src/test/java/org/sonar/server/rule/RubyRuleServiceTest.java");
+
+    action
+      .createParam(TEST_UUID)
+      .setDescription("Test UUID")
+      .setExampleValue("c526ef20-131b-4486-9357-063fa64b5079");
+
+    action
+      .createParam(SOURCE_FILE_UUID)
```

Comment:
Why have it ? TEST_FILE_KEY is provided only to be in an IT. It would have been the TEST_FILE_UUID only otherwise.